File: src/Api/Model/FileSyncMessage.cs
namespace Api.Model;

public abstract class FileSyncMessage
{
    public abstract FileSyncMessageType MessageType { get; }
}

public class FileUpdatedMessage : FileSyncMessage
{
    public override FileSyncMessageType MessageType => FileSyncMessageType.UpdateTime;
    public DateTime Time { get; }
    public string SenderClientId { get; }

    public FileUpdatedMessage(DateTime time, string senderClientId)
    {
        Time = time;
        SenderClientId = senderClientId;
    }
}

public class ClientIdMessage : FileSyncMessage
{
    public override FileSyncMessageType MessageType => FileSyncMessageType.ClientId;
    public string ClientId { get; }

    public ClientIdMessage(string clientId)
    {
        ClientId = clientId;
    }
}

public enum FileSyncMessageType
{
    UpdateTime,
    ClientId
}

File: src/Web/src/model/file-sync-message.ts
export enum FileSyncMessageType {
    UpdateTime = 0,
    ClientId = 1,
}

export interface FileUpdatedMessage {
    messageType: FileSyncMessageType.UpdateTime;
    time: string | Date;
    senderClientId: string;
}

export interface ClientIdMessage {
    messageType: FileSyncMessageType.ClientId;
    clientId: string;
}

export type FileSyncMessage = FileUpdatedMessage | ClientIdMessage;
